### PR TITLE
Connect asset database to external trackers

### PR DESCRIPTION
**Pull request by benizzio**
This pull request introduces a new external asset search API endpoint and adds infrastructure for running external integration tests, along with some internal refactoring and database schema changes. The changes are grouped below by theme.

**External Asset Search API:**

* Added a new GET endpoint `/api/external-asset` to the REST API, allowing clients to search for external assets by query. This includes new data transfer structures (`ExternalAssetDTS`, `ExternalAssetSearchQueryDTS`) and mapping functions in `asset_rest_model.go`. [[1]](diffhunk://#diff-9736d79c94a4ce37f1df1e83772cde78378d03b390d47c21c121f251e345a8c7R37-R41) [[2]](diffhunk://#diff-9736d79c94a4ce37f1df1e83772cde78378d03b390d47c21c121f251e345a8c7R123-R143) [[3]](diffhunk://#diff-ffeaf1a5133b095ff9c7ea71013b6adc8a6b41b5f93db7370d64e1b4a54c462bR18-R39) [[4]](diffhunk://#diff-ffeaf1a5133b095ff9c7ea71013b6adc8a6b41b5f93db7370d64e1b4a54c462bR90-R135) [[5]](diffhunk://#diff-403901ce0ee26285d6ee1aa5f44f1af8af3ca855f893d984d5e4923e3d1f7850R8-R9)
* Updated documentation in `AGENTS.md` to clarify that `inttest` should not access external APIs and introduced `extinttest` for tests that do.

**External Integration Test Infrastructure:**

* Added a new GitHub Actions workflow job (`ext-test`) to run Go-based external integration tests, along with a corresponding `test-ext` target in the `Makefile`. Includes a setup guide in `TODO-ext-test-setup.md` for configuring required CI checks and environments. [[1]](diffhunk://#diff-1db27d93186e46d3b441ece35801b244db8ee144ff1405ca27a163bfe878957fR43-R71) [[2]](diffhunk://#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R16-R19) [[3]](diffhunk://#diff-f4d3d1f6447da6e0a06d2dfb849556a158407fb0718a3f70a12cb380e7e273edR1-R18)

**Database and Domain Model Changes:**

* Added an `external_data` JSONB column to the `asset` table to store external asset information. ([src/main/flyway/sql/migration-[14]-asset_external_data.sqlR1](diffhunk://#diff-2eb1786362ca310ba8685970ed08f783bcb6d1687ca39336a38da89f8f7c2a07R1))
* Updated the `Asset` domain model to include an `ExternalData` field.

**Codebase Refactoring:**

* Refactored the `AllocationStructure` scan/value logic to use shared helpers from `sqlext`, and moved the `HierarchicalId` type and related logic into a new file (`allocation_hierarchical_id.go`). Extracted the `AllocationRepository` interface into its own file (`allocation_repository.go`). [[1]](diffhunk://#diff-c59f73c5a804b90ff720b92fbf3e89d295878094020ef864d93732940c6f65fdL5-R6) [[2]](diffhunk://#diff-c59f73c5a804b90ff720b92fbf3e89d295878094020ef864d93732940c6f65fdL39-R41) [[3]](diffhunk://#diff-86dfc3f6c4a732d39e29960edd32f8f9154e30ade390c2a4a7c14c9a9c763600R1-R80) [[4]](diffhunk://#diff-d6633ed94e5a196854a9140c43bdf4191041d37836caf865634ec9bc4a312da2R1-R9)

**Dependency Updates:**

* Added new Go module dependencies required for external integration tests and related tooling in `go.work.sum`. [[1]](diffhunk://#diff-20cee3a725d1c5d2c595ff5372e6593d3bcaa7aebd8319866f65e90c6924f247R9) [[2]](diffhunk://#diff-20cee3a725d1c5d2c595ff5372e6593d3bcaa7aebd8319866f65e90c6924f247R22) [[3]](diffhunk://#diff-20cee3a725d1c5d2c595ff5372e6593d3bcaa7aebd8319866f65e90c6924f247R32-R42) [[4]](diffhunk://#diff-20cee3a725d1c5d2c595ff5372e6593d3bcaa7aebd8319866f65e90c6924f247R59-R80)

## Summary by CodeRabbit

* **New Features**
  * Added an external asset search endpoint (query external providers like Yahoo Finance) and ability to fetch external asset prices.

* **Database**
  * Assets can now store external metadata for enriched asset records.

* **Tests / CI**
  * Live external integration tests added and a CI job to run them.

* **Documentation**
  * Guidance added for configuring and gating external integration tests in CI.

### CODE REVIEW

File: src/main/go/api/rest/model/asset_rest_model.go [hunk 1]
```diff
@@ -65,3 +87,49 @@ func MapToAssets(assetsDTS []*AssetDTS) []*domain.Asset {
...
+// Parameters:
+//   - externalAssets: the slice of domain ExternalAsset pointers to map
+//
+// Returns:
+//   - []*ExternalAssetDTS: the mapped slice of REST DTSs
+//
+// Authored by: GitHub Copilot (claude-opus-4.6)
+func MapToExternalAssetDTSs(externalAssets []*domain.ExternalAsset) []*ExternalAssetDTS {
+	var externalAssetDTSs = make([]*ExternalAssetDTS, len(externalAssets))
+	for i, externalAsset := range externalAssets {
```
Comment (benizzio) [inline src/main/go/api/rest/model/asset_rest_model.go]: rename `i` to `index` to keep the meaningful name standards

File: src/main/go/inttest/asset_test_utils.go [hunk 2]
```diff
@@ -91,3 +93,27 @@ func putAsset(t *testing.T, putAssetJSON string) *http.Response {
...
+	var response, err = http.Get(requestURL)
+	assert.NoError(t, err)
+
+	var responseBodyBytes, readErr = io.ReadAll(response.Body)
+	assert.NoError(t, readErr)
+
+	err = response.Body.Close()
+	assert.NoError(t, err)
+
+	return response, string(responseBodyBytes)
```
Comment (benizzio) [inline src/main/go/inttest/asset_test_utils.go]: Instead of returning the response here, return just the status code

File: src/main/go/langext/golang_ext_util_concurrent.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+func FlatMapConcurrently[I any, R any](inputs []I, sliceFromItem SliceFromItem[I, R]) ([]R, error) {
+
+	var resultChannel = make(chan concurrentSliceResult[R], len(inputs))
+
+	for _, input := range inputs {
+		go executeSliceFromItemConcurrently(input, sliceFromItem, resultChannel)
+	}
+
+	return collectConcurrentResults(resultChannel, len(inputs))
+}
```
[the remark this answers is not in the document]
Comment (benizzio) [reply src/main/go/langext/golang_ext_util_concurrent.go:86]: Added `src/main/go/langext/golang_ext_util_concurrent_test.go` covering empty input, result aggregation, and first-error propagation/cancellation for the new helper. Verified with `go test ./...`, `go test -count=1 -tags=extinttest ./extinttest/...`, and `make lint`.

File: .github/workflows/tests.yml [hunk 4]
```diff
@@ -40,3 +40,32 @@ jobs:
...
+        id: go-version
+        run: |
+          GO_VERSION=$(grep -E "^go [0-9]+\.[0-9]+" src/main/go/go.mod | awk '{print $2}')
+          echo "GO_VERSION=$GO_VERSION" >> $GITHUB_OUTPUT
+
+      - name: Setup Go
+        uses: actions/setup-go@v4
+        with:
+          go-version: ${{ steps.go-version.outputs.GO_VERSION }}
+
```
[the remark this answers is not in the document]
Comment (benizzio) [reply .github/workflows/tests.yml]: Bumped both workflow jobs to `actions/checkout@v5` and `actions/setup-go@v5`. `make lint` is clean after the workflow update.

File: src/main/go/domain/infra/integration/asset_integration_client_yahoo.go [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+		return "", fmt.Errorf("error parsing Yahoo Finance chart URL %s: %w", chartURL, err)
+	}
+
+	var queryParams = parsedURL.Query()
+	queryParams.Set("interval", "1d")
+	queryParams.Set("events", "history")
+
+	parsedURL.RawQuery = queryParams.Encode()
+
+	return parsedURL.String(), nil
```
[the remark this answers is not in the document]
Comment (benizzio) [reply src/main/go/domain/infra/integration/asset_integration_client_yahoo.go:174]: Applied this in `buildQuoteAssetLastClosePriceURL` by parsing the base URL and appending `url.PathEscape(ticker)` to the path before query parameters are encoded.

File: src/main/go/domain/infra/integration/asset_integration_client_yahoo_model.go [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+	Quotes []YahooFinanceSearchQuoteDTS `json:"quotes"`
+}
+
+// YahooFinanceChartQuoteIndicatorDTS represents the OHLCV quote indicator data
+// from the Yahoo Finance chart API response.
+//
+// Authored by: GitHub Copilot (claude-opus-4.6)
+type YahooFinanceChartQuoteIndicatorDTS struct {
+	Close []float64 `json:"close"`
+}
```
[the remark this answers is not in the document]
Comment (benizzio) [reply src/main/go/domain/infra/integration/asset_integration_client_yahoo_model.go:34]: Changed `YahooFinanceChartQuoteIndicatorDTS.Close` to `[]*float64` and updated the last-close extraction path to walk backward, skip `nil` values, and pair the selected close with the matching timestamp index. The external integration test was updated accordingly.

File: src/main/go/domain/infra/integration/asset_integration_client_yahoo.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (benizzio) [reply src/main/go/domain/infra/integration/asset_integration_client_yahoo.go:174]: Updated `buildQuoteAssetLastClosePriceURL` to parse the base URL first and append `url.PathEscape(ticker)` to the path instead of concatenating the raw ticker string.

File: src/main/go/inttest/asset_test_utils.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (benizzio) [reply src/main/go/inttest/asset_test_utils.go]: Updated `getExternalAssets` to return only `(statusCode, body)` instead of `*http.Response`, and adjusted callers accordingly. I also switched the transport/body-read checks in that helper to `require.NoError(...)` so failures stop immediately.

File: src/main/go/infra/util/http/httpclient/http_client.go [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+
+	for _, option := range options {
+		option(request)
+	}
+
+	var client = &http.Client{}
+	response, err := client.Do(request)
+	if err != nil {
+		return nil, err
+	}
```
[the remark this answers is not in the document]
Comment (benizzio) [reply src/main/go/infra/util/http/httpclient/http_client.go]: Changed `ExecuteGet` and `ExecuteGetJSON` to accept request context, build requests with `http.NewRequestWithContext`, and use a default `http.Client{Timeout: 10 * time.Second}`. Verified with `go test ./...`, `go test -count=1 -tags=extinttest ./extinttest/...`, and `make lint`.

File: src/main/go/domain/service/asset_domain_service.go [hunk 8]
```diff
@@ -55,8 +59,48 @@ func (service *AssetDomService) InsertMappedAssetsInTransaction(
...
+
+	var integrationServices = collectIntegrationServices(service.assetIntegrationServicesPerSource)
+
+	var searchAssetsOnService = func(
+		integrationService domain.AssetIntegrationService,
+	) ([]*domain.ExternalAsset, error) {
+		return integrationService.SearchAssets(query)
+	}
+
+	return langext.FlatMapConcurrently(integrationServices, searchAssetsOnService)
```
[the remark this answers is not in the document]
Comment (benizzio) [reply src/main/go/domain/service/asset_domain_service.go]: Threaded request context through `AssetDomService.SearchExternalAssets`, `AssetIntegrationService.SearchAssets`, the Yahoo search client/service, and the new `langext.FlatMapConcurrentlyCtx` path so provider fan-out work can stop on request cancellation or first error.

File: src/main/go/infra/gin/gin_validation_query.go [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+		allErrorMessages = append(allErrorMessages, validationErrors...)
+	}
+
+	var deepValidationErrors = validation.DeepValidate(bindingTarget)
+	allErrorMessages = append(allErrorMessages, deepValidationErrors...)
+
+	if len(allErrorMessages) > 0 {
+		sendValidationErrorResponse(context, allErrorMessages)
+		return false, nil
```
[the remark this answers is not in the document]
Comment (benizzio) [reply src/main/go/infra/gin/gin_validation_query.go:33]: Updated `BindAndValidateQueryWithInvalidResponse` so unrecognized bind/parse failures stay on the validation response path with the generic fallback message `malformed or invalid query parameters`. Added `src/main/go/infra/gin/gin_validation_query_test.go` to cover that case.

File: src/main/go/infra/rdbms/sqlext/sqlext_null_time.go [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+	"database/sql"
+	"time"
+)
+
+type NullTime sql.NullTime
+
+func (nullTime *NullTime) ToTimeReference() *time.Time {
+	if nullTime.Valid {
+		return &nullTime.Time
+	}
```
[the remark this answers is not in the document]
Comment (benizzio) [reply src/main/go/infra/rdbms/sqlext/sqlext_null_time.go:13]: Changed `NullTime.ToTimeReference()` to return `nil` for invalid values and a pointer to a copied `time.Time` value for valid ones, so callers do not receive a pointer to receiver-owned storage.

File: src/main/go/inttest/infra/yahoo_finance_mock.go [hunk 11]
```diff
@@ -0,0 +1,69 @@
+package infra
```
[the remark this answers is not in the document]
Comment (benizzio) [reply src/main/go/inttest/infra/yahoo_finance_mock.go:5]: Added the package-level documentation block above `package infra` in `yahoo_finance_mock.go`, including the purpose of the shared Yahoo Finance test mock and the authoring line.

File: src/main/go/infra/rdbms/sqlext/sqlext_utils.go [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+	if value == nil {
+		return nil
+	}
+
+	bytes, ok := value.([]byte)
+	if !ok {
+		return fmt.Errorf("scanned value is incompatible (not a []byte): %#v", value)
+	}
+
+	return json.Unmarshal(bytes, target)
```
[the remark this answers is not in the document]
Comment (benizzio) [reply src/main/go/infra/rdbms/sqlext/sqlext_utils.go:24]: Updated `ScanJsonColumn` to zero the destination before returning on SQL `NULL`, so reused scan targets do not retain stale decoded JSON data.

File: src/main/go/api/rest/model/asset_rest_model.go (hunk 1 above)
Comment (benizzio) [reply src/main/go/api/rest/model/asset_rest_model.go]: Renamed the loop variable from `i` to `index` in the external asset mapping loops to match the naming standard.

File: src/main/go/inttest/infra/yahoo_finance_mock.go [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+// SetupYahooFinanceMockTest resets the shared Yahoo Finance mock server for the given test and
+// registers cleanup that verifies all expectations were met and clears state afterwards.
+//
+// Authored by: GitHub Copilot
+func SetupYahooFinanceMockTest(t *testing.T) *httpmock.Server {
+	t.Helper()
+
+	var mockServer = GetYahooFinanceMockServer()
+	mockServer.WithTest(t)
+	resetYahooFinanceMockServer(mockServer)
```
[the remark this answers is not in the document]
Comment (benizzio) [reply src/main/go/inttest/infra/yahoo_finance_mock.go:54]: Added the explicit nil guard in `SetupYahooFinanceMockTest` so missing suite bootstrap fails fast with a clear test error instead of a nil dereference.

File: src/main/go/infra/util/http/httpclient/http_client.go [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+
+	if response.StatusCode != http.StatusOK {
+		CloseResponseBody(response)
+		return nil, fmt.Errorf("HTTP GET request to %s returned status %d", requestURL, response.StatusCode)
+	}
+
+	return response, nil
+}
```
[the remark this answers is not in the document]
Comment (benizzio) [reply src/main/go/infra/util/http/httpclient/http_client.go]: Implemented the context-aware version of this suggestion: `ExecuteGet` and `ExecuteGetJSON` now accept request context, create requests with `http.NewRequestWithContext`, and also use a default `10s` client timeout as a baseline safeguard.

File: src/main/go/inttest/asset_test_utils.go (hunk 2 above)
Comment (benizzio) [reply src/main/go/inttest/asset_test_utils.go]: Changed the helper signature to return only `(statusCode, body)` and updated all integration test call sites to use the returned status code directly.

File: src/main/go/langext/golang_ext_util_concurrent.go [hunk 15]
```diff
@@ -0,0 +1,287 @@
...
+) ([]R, error) {
+
+	if len(inputs) == 0 {
+		return make([]R, 0), nil
+	}
+
+	var workContext, cancel = context.WithCancel(ctx)
+	defer cancel()
+
+	var inputChannel = buildConcurrentInputChannel(ctx, inputs)
```
[the remark this answers is not in the document]
Comment (benizzio) [reply src/main/go/langext/golang_ext_util_concurrent.go]: Changed `FlatMapConcurrentlyCtx` to build the producer channel with `workContext` instead of the parent `ctx`, so the producer observes the same cancellation signal as the workers. I also added a regression test in `src/main/go/langext/golang_ext_util_concurrent_test.go` covering the `FlatMapConcurrently` error path to guard against producer goroutine leaks.

File: src/main/go/domain/allocation_hierarchical_id.go [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+
+	var length = len(hierarchicalId)
+	var lastIndex = length - 1
+
+	if lastIndex == 0 {
+		return true
+	}
+
+	return hierarchicalId[lastIndex] != nil && hierarchicalId[lastIndex-1] == nil
+}
```
[the remark this answers is not in the document]
Comment (benizzio) [reply src/main/go/domain/allocation_hierarchical_id.go:51]: Added the empty-slice guard in `HierarchicalId.IsTopLevel()` so it now returns `false` before computing `lastIndex`. I also added regression coverage in `src/main/go/domain/hierarchical_id_test.go` for both `IsTopLevel()` and `ParentLevelId()` on an empty `HierarchicalId`.

File: src/main/go/infra/util/http/httpclient/http_client.go [hunk 17]
```diff
@@ -0,0 +1,198 @@
...
+	"github.com/golang/glog"
+)
+
+const defaultTimeout = 10 * time.Second
+
+// RequestOption is a functional option that configures an HTTP request before execution.
+// Used with ExecuteGet and ExecuteGetJSON to customize request headers, parameters, or
+// other properties without modifying the function signatures for each new requirement.
+//
+// Example:
```
Comment (benizzio) [inline src/main/go/infra/util/http/httpclient/http_client.go]: Types should not have examples in comment docs, only functions

**Review by benizzio — COMMENTED** [review]
Also fix all tests touched in this PR that do assertions via `if` and `fatal`. In the project assertions should be done by `github.com/stretchr/testify/assert` or `github.com/stretchr/testify/require`

File: src/main/go/domain/allocation_hierarchical_id.go [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+// String returns the hierarchical identifier as a single string using
+// HierarchicalIdLevelSeparator between non-nil levels.
+func (hierarchicalId HierarchicalId) String() string {
+	var result strings.Builder
+	for index, level := range hierarchicalId {
+		if level != nil {
+			result.WriteString(*level)
+			if index < len(hierarchicalId)-1 {
+				result.WriteString(HierarchicalIdLevelSeparator)
+			}
```
[the remark this answers is not in the document]
Comment (benizzio) [reply src/main/go/domain/allocation_hierarchical_id.go:21]: To avoid bloating the PR we will skip this comment at this time

File: src/main/go/domain/allocation_hierarchical_id.go [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+		return false
+	}
+
+	var lastIndex = length - 1
+
+	if lastIndex == 0 {
+		return true
+	}
+
+	return hierarchicalId[lastIndex] != nil && hierarchicalId[lastIndex-1] == nil
```
[the remark this answers is not in the document]
Comment (benizzio) [reply src/main/go/domain/allocation_hierarchical_id.go:50]: To avoid bloating the PR we will skip this comment at this time
